### ISSUE TITLE
Incorrect and missing tautological warnings with consteval if

### ISSUE THREAD

**Issue by AaronBallman** [llvm/llvm-project#51567]
|  |  |
| --- | --- |
| Bugzilla Link | [52225](https://llvm.org/bz52225) |
| Version | trunk |
| OS | Windows NT |
| CC | @zygoloid |

## Extended Description 
The following consteval if statements produce tautological warnings that don't seem correct in all cases.
```
constexpr void f() {
  if consteval {
    if consteval { // Correct diagnostic
    }
    if !consteval { // Incorrect diagnostic
    }
  }

  if !consteval {
    if consteval { // Missing diagnostic
    }
    if !consteval { // Missing diagnostic
    }
  }
}
```
https://godbolt.org/z/6KjW4v5dn

**Comment by tbaederr** [comment]
The problems with the `!consteval` block are happening because no expression evaluation context is being pushed for it in `ParseStmt.cpp`:
https://github.com/llvm/llvm-project/blob/ce2345d8c14b2c8b9a420b8ca53dc1ea5d43dc6a/clang/lib/Parse/ParseStmt.cpp#L1458-L1461

So the later call to `isImmediateFunctionContext()` returns `false` here:

https://github.com/llvm/llvm-project/blob/ce2345d8c14b2c8b9a420b8ca53dc1ea5d43dc6a/clang/lib/Sema/SemaStmt.cpp#L924-L938

and no diagnostic is emitted.

**Comment by Izaron** [comment]
The main problem is `if ! consteval`, because now we can't say if we are inside it. This could be solved by a boolean flag at `Sema` class indicating if we are in a scope within `if ! consteval` (IMO a better way); or by another Context type (IMO a worser way, if ever possible).

Both ways will induct memory footprint, and make code harder. I wonder if it's a reasonable price for detecting errors that very few programmers will ever make.
